soccer/ui/MainWindow.cpp

```diff
@@ -430,32 +457,19 @@ void MainWindow::updateViews() {
 
     // Check if any debug layers have been added
     // (layers should never be removed)
-    const std::shared_ptr<LogFrame> liveFrame =
-        _processor->logger().lastFrame();
-    if (liveFrame &&
-        liveFrame->debug_layers_size() > _ui.debugLayers->count()) {
-        // Add the missing layers and turn them on
-        for (int i = _ui.debugLayers->count();
-             i < liveFrame->debug_layers_size(); ++i) {
-            const QString name =
-                QString::fromStdString(liveFrame->debug_layers(i));
-            bool enabled =
-                !std::any_of(defaultHiddenLayers.begin(),
-                             defaultHiddenLayers.end(),
-                             [&](QString string) { return string == name; });
-            addLayer(i, name, enabled);
-        }
-
-        _ui.debugLayers->sortItems();
+    if (live_frame) {
+        updateDebugLayers(*live_frame);
     }
 
     // Get the frame at the log playback time
-    const std::shared_ptr<LogFrame> currentFrame = _history[0];
+    const std::shared_ptr<LogFrame> currentFrame = _history.back();
 
     if (currentFrame) {
+        _ui.logTree->message(*currentFrame);
```

Comment:
This call is taking a long time. What exactly does the protobuf tree do? Can we do this in a separate thread or something?

Reply:
It's using a lot of reflection and updating GUI, which is probably why it's so slow. I could make this only update every few frames - wouldn't really _fix_ the problem, but would at least make it better. Is it taking a substantial portion of frame times?